pkg/service/controller/servingcert/secret_creating_controller.go

```diff
@@ -339,9 +378,73 @@ func (sc *ServiceServingCertController) requiresCertGeneration(service *v1.Servi
 	if service.Annotations[ServingCertCreatedByAnnotation] == sc.ca.Config.Certs[0].Subject.CommonName {
 		return false
 	}
-	// TODO: use the lister here
-	if _, exists, _ := sc.secretCache.GetByKey(service.Namespace + "/" + secretName); !exists {
+	_, err := sc.secretLister.Secrets(service.Namespace).Get(secretName)
+	if kapierrors.IsNotFound(err) {
 		return true
 	}
+	if err != nil {
+		glog.Errorf("Unable to get the secret %s/%s: %v", service.Namespace, secretName, err)
+		return false
+	}
 	return true
 }
+
+// updateSecretOwnerReference updates the service secrets that does not have the
+// owner reference set to the service they belongs to.
+func (sc *ServiceServingCertController) updateSecretOwnerReference(service *v1.Service) (bool, error) {
+	secretName := service.Annotations[ServingCertSecretAnnotation]
+	if len(secretName) == 0 {
+		return false, nil
+	}
+	secret, err := sc.secretLister.Secrets(service.Namespace).Get(secretName)
+	if kapierrors.IsNotFound(err) {
+		return false, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	t, err := kapi.Scheme.DeepCopy(secret)
+	if err != nil {
+		return false, err
+	}
+	secretCopy := t.(*v1.Secret)
+	refUpdated, err := addOwnerReference(secretCopy, service)
+	if err != nil {
+		return false, err
+	}
+	if !refUpdated {
+		return false, nil
+	}
+	_, err = sc.secretClient.Secrets(secretCopy.Namespace).Update(secretCopy)
+	return true, err
+}
+
+// addOwnerReference adds an owner reference to a secret pointing to the service
+// that the secret was created for.
+func addOwnerReference(secret *v1.Secret, service *v1.Service) (bool, error) {
+	svcGvk, isUnversioned, err := kapi.Scheme.ObjectKind(service)
+	if err != nil {
+		return false, err
+	}
+	if isUnversioned {
+		return false, fmt.Errorf("unexpected unversioned object: %v", service)
+	}
+	hasOwnerRef := false
+	for _, r := range secret.OwnerReferences {
+		if r.UID == service.UID && r.Kind == svcGvk.Kind {
+			hasOwnerRef = true
+			break
+		}
+	}
+	if hasOwnerRef {
+		return false, nil
+	}
+	secret.OwnerReferences = append(secret.OwnerReferences, metav1.OwnerReference{
+		APIVersion: svcGvk.GroupVersion().String(),
```

Comment:
@deads2k need double-check, is this controllerRef since the controller manage these secrets?

Reply:
> @deads2k need double-check, is this controllerRef since the controller manage these secrets?

`controllerRef` is a field that should not be.  I wouldn't set it.  What possible purpose could it serve, since multiple controllers can touch a single object?  Even worse, validation forces this to be a single ownerref, so setting it is just asking for failures.  To cap it all off, the value can't logically be used to unless the controllers *already* negotiate with each other, so the field was a re-use of something which shouldn't exist.  And just to be sure the field was wrong, it doesn't handle cases where the controlling entity exists outside the API (say it was driven off a master-config).